File: ralph-dashboard/src/server/jev-usage.ts
import { promises as fs } from 'node:fs';
import { dirname, join } from 'node:path';

/**
 * Jev (TypeSafe AI) is an HTTP adapter, not a CLI runtime, so its usage is not
 * part of the runtime token/cache metrics. The transport appends one line per
 * successful call to <workspace_root>/jev/usage.jsonl; this reads and sums them.
 */

const DEFAULT_INPUT_USD_PER_MTOK = 0.042;
const DEFAULT_OUTPUT_USD_PER_MTOK = 0;
const DAY_MS = 24 * 60 * 60 * 1000;

export interface JevUsageQuestionSetRow {
  question_set_id: string;
  calls: number;
  input_tokens: number;
  output_tokens: number;
}

export interface JevUsageResponse {
  /** True only when at least one live Jev call was recorded in scope. */
  enabled: boolean;
  calls: number;
  calls_measured: number;
  calls_unavailable: number;
  input_tokens: number;
  output_tokens: number;
  estimated_usd: number;
  by_question_set: JevUsageQuestionSetRow[];
}

function nonNegInt(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? Math.max(0, Math.floor(value)) : 0;
}

function rate(name: string, fallback: number): number {
  const raw = process.env[name]?.trim();
  if (!raw) {
    return fallback;
  }
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : fallback;
}

function scopeBounds(dateFrom?: string, dateTo?: string): { from: number; to: number } {
  const from = dateFrom ? Date.parse(dateFrom) : NaN;
  let to = dateTo ? Date.parse(dateTo) : NaN;
  // A bare YYYY-MM-DD upper bound means the whole day.
  if (dateTo && /^\d{4}-\d{2}-\d{2}$/.test(dateTo) && Number.isFinite(to)) {
    to += DAY_MS - 1;
  }
  return {
    from: Number.isFinite(from) ? from : Number.NEGATIVE_INFINITY,
    to: Number.isFinite(to) ? to : Number.POSITIVE_INFINITY,
  };
}

export async function collectJevUsage(options: {
  /** Ralph logs roots (<workspace_root>/logs); the Jev state dir is a sibling. */
  logsRoots: string[];
  dateFrom?: string;
  dateTo?: string;
}): Promise<JevUsageResponse> {
  const { from, to } = scopeBounds(options.dateFrom, options.dateTo);
  const bySet = new Map<string, JevUsageQuestionSetRow>();
  let calls = 0;
  let measured = 0;
  let unavailable = 0;
  let inputTokens = 0;
  let outputTokens = 0;

  const seen = new Set<string>();
  for (const logsRoot of options.logsRoots) {
    const file = join(dirname(logsRoot), 'jev', 'usage.jsonl');
    if (seen.has(file)) {
      continue;
    }
    seen.add(file);
    let raw: string;
    try {
      raw = await fs.readFile(file, 'utf8');
    } catch {
      continue;
    }
    for (const line of raw.split('\n')) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }
      let rec: Record<string, unknown>;
      try {
        const parsed: unknown = JSON.parse(trimmed);
        if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
          continue;
        }
        rec = parsed as Record<string, unknown>;
      } catch {
        continue;
      }
      // Offline fixture replays are test traffic, not real spend.
      if (rec['transport'] === 'fixture') {
        continue;
      }
      if (from !== Number.NEGATIVE_INFINITY || to !== Number.POSITIVE_INFINITY) {
        const ts = typeof rec['timestamp'] === 'string' ? Date.parse(rec['timestamp']) : NaN;
        if (!Number.isFinite(ts) || ts < from || ts > to) {
          continue;
        }
      }
      const input = nonNegInt(rec['input_tokens']);
      const output = nonNegInt(rec['output_tokens']);
      calls += 1;
      inputTokens += input;
      outputTokens += output;
      if (rec['usageSource'] === 'measured') {
        measured += 1;
      } else {
        unavailable += 1;
      }
      const id =
        typeof rec['questionSetId'] === 'string' && rec['questionSetId']
          ? rec['questionSetId']
          : '(unnamed)';
      const row = bySet.get(id) ?? { question_set_id: id, calls: 0, input_tokens: 0, output_tokens: 0 };
      row.calls += 1;
      row.input_tokens += input;
      row.output_tokens += output;
      bySet.set(id, row);
    }
  }

  const usd =
    (inputTokens * rate('RALPH_JEV_INPUT_USD_PER_MTOK', DEFAULT_INPUT_USD_PER_MTOK)) / 1_000_000 +
    (outputTokens * rate('RALPH_JEV_OUTPUT_USD_PER_MTOK', DEFAULT_OUTPUT_USD_PER_MTOK)) / 1_000_000;

  return {
    enabled: calls > 0,
    calls,
    calls_measured: measured,
    calls_unavailable: unavailable,
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    estimated_usd: Math.round(usd * 1_000_000) / 1_000_000,
    by_question_set: [...bySet.values()].sort((a, b) => b.calls - a.calls),
  };
}

File: bundle/.ralph/python/jev_usage.py
#!/usr/bin/env python3
"""Aggregate Jev (TypeSafe AI) per-call usage from <jev_state_dir>/usage.jsonl.

Jev is an HTTP adapter, not a CLI runtime, so its usage is kept apart from the
runtime/cache buckets in invocation-usage.json. Records are written by the
transport (jev_client.record_usage / jev_record_usage in jev-client.sh).

Usage:
  jev_usage.py [--state-dir DIR | --file PATH] [--plan-key KEY]
               [--include-fixture] [--omit-empty] [--format text|json]
  jev_usage.py --merge-json [same options]   # reads a JSON object on stdin,
                                             # adds a "jev" key when there is
                                             # usage, and prints it.

Fixture-transport lines (offline tests) are ignored unless --include-fixture.
Cost is an estimate: rates come from RALPH_JEV_INPUT_USD_PER_MTOK (default
0.042) and RALPH_JEV_OUTPUT_USD_PER_MTOK (default 0, output tokens are free);
see docs/TOOLING.md "Cost and latency".
"""
from __future__ import annotations

import argparse
import json
import os
import sys
from typing import Any, Iterable, Mapping, Optional

DEFAULT_INPUT_USD_PER_MTOK = 0.042
DEFAULT_OUTPUT_USD_PER_MTOK = 0.0


def _int(value: Any) -> int:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return 0
    return max(0, int(value))


def _rate(name: str, default: Optional[float]) -> Optional[float]:
    raw = os.environ.get(name, "").strip()
    if not raw:
        return default
    try:
        val = float(raw)
    except ValueError:
        return default
    return val if val >= 0 else default


def state_dir() -> str:
    explicit = os.environ.get("RALPH_JEV_STATE_DIR", "").strip()
    if explicit:
        return explicit.rstrip("/")
    root = os.environ.get("RALPH_PLAN_WORKSPACE_ROOT", "").strip() or os.path.join(
        os.getcwd(), ".ralph-workspace"
    )
    return os.path.join(root.rstrip("/"), "jev")


def read_records(path: str) -> Iterable[Mapping[str, Any]]:
    try:
        handle = open(path, "r", encoding="utf-8")
    except OSError:
        return
    with handle:
        for line in handle:
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue
            if isinstance(obj, dict):
                yield obj


def aggregate(
    records: Iterable[Mapping[str, Any]],
    plan_key: str = "",
    include_fixture: bool = False,
) -> dict[str, Any]:
    by_set: dict[str, dict[str, Any]] = {}
    by_model: dict[str, dict[str, int]] = {}
    total = {
        "calls": 0,
        "input_tokens": 0,
        "output_tokens": 0,
        "calls_measured": 0,
        "calls_unavailable": 0,
    }
    for rec in records:
        if not include_fixture and rec.get("transport") == "fixture":
            continue
        if plan_key and rec.get("planKey") != plan_key:
            continue
        in_tok = _int(rec.get("input_tokens"))
        out_tok = _int(rec.get("output_tokens"))
        measured = rec.get("usageSource") == "measured"
        total["calls"] += 1
        total["input_tokens"] += in_tok
        total["output_tokens"] += out_tok
        total["calls_measured" if measured else "calls_unavailable"] += 1

        qs = str(rec.get("questionSetId") or "(unnamed)")
        row = by_set.setdefault(
            qs, {"calls": 0, "input_tokens": 0, "output_tokens": 0}
        )
        row["calls"] += 1
        row["input_tokens"] += in_tok
        row["output_tokens"] += out_tok

        model = str(rec.get("model") or "(unresolved)")
        mrow = by_model.setdefault(
            model, {"calls": 0, "input_tokens": 0, "output_tokens": 0}
        )
        mrow["calls"] += 1
        mrow["input_tokens"] += in_tok
        mrow["output_tokens"] += out_tok

    in_rate = _rate("RALPH_JEV_INPUT_USD_PER_MTOK", DEFAULT_INPUT_USD_PER_MTOK)
    out_rate = _rate("RALPH_JEV_OUTPUT_USD_PER_MTOK", DEFAULT_OUTPUT_USD_PER_MTOK)
    cost: Optional[dict[str, Any]] = None
    if total["calls"]:
        usd = 0.0
        if in_rate is not None:
            usd += total["input_tokens"] * in_rate / 1_000_000
        if out_rate is not None:
            usd += total["output_tokens"] * out_rate / 1_000_000
        cost = {
            "estimated_usd": round(usd, 6),
            "input_rate_usd_per_mtok": in_rate,
            "output_rate_usd_per_mtok": out_rate,
            "note": "estimated",
        }
    return {
        "kind": "jev_usage",
        "schema_version": 1,
        **total,
        "by_question_set": by_set,
        "by_model": by_model,
        "cost": cost,
    }


def render_text(summary: Mapping[str, Any]) -> str:
    lines = ["Jev (TypeSafe AI) usage"]
    lines.append(
        "  calls: %d (measured %d, usage unavailable %d)"
        % (summary["calls"], summary["calls_measured"], summary["calls_unavailable"])
    )
    lines.append(
        "  tokens: input %d, output %d"
        % (summary["input_tokens"], summary["output_tokens"])
    )
    cost = summary.get("cost")
    if isinstance(cost, Mapping):
        lines.append("  cost: ~$%.6f (%s)" % (cost["estimated_usd"], cost["note"]))
    sets = summary.get("by_question_set") or {}
    if sets:
        lines.append("  by question set:")
        for name in sorted(sets, key=lambda k: -sets[k]["calls"]):
            row = sets[name]
            lines.append(
                "    %-32s calls %-5d in %-8d out %d"
                % (name, row["calls"], row["input_tokens"], row["output_tokens"])
            )
    return "\n".join(lines)


def main(argv: Optional[list[str]] = None) -> int:
    ap = argparse.ArgumentParser(description=__doc__.split("\n")[0])
    ap.add_argument("--state-dir", default="")
    ap.add_argument("--file", default="")
    ap.add_argument("--plan-key", default="")
    ap.add_argument("--include-fixture", action="store_true")
    ap.add_argument("--format", choices=("text", "json"), default="text")
    ap.add_argument("--merge-json", action="store_true")
    ap.add_argument(
        "--omit-empty",
        action="store_true",
        help="print nothing (text) when there are no recorded calls",
    )
    args = ap.parse_args(argv)

    path = args.file or os.path.join(args.state_dir or state_dir(), "usage.jsonl")
    summary = aggregate(read_records(path), args.plan_key, args.include_fixture)

    if args.merge_json:
        raw = sys.stdin.read()
        try:
            base = json.loads(raw)
        except json.JSONDecodeError:
            sys.stdout.write(raw)
            return 0
        if isinstance(base, dict) and summary["calls"]:
            base["jev"] = summary
        json.dump(base, sys.stdout, indent=2)
        sys.stdout.write("\n")
        return 0

    if args.format == "json":
        json.dump(summary, sys.stdout, indent=2)
        sys.stdout.write("\n")
    elif summary["calls"]:
        print(render_text(summary))
    elif not args.omit_empty:
        print("Jev (TypeSafe AI) usage: no recorded calls")
    return 0


if __name__ == "__main__":
    sys.exit(main())
